### ISSUE TITLE
shadowed variable can lead to runtime error in generated JS

### ISSUE THREAD

**Issue by jvoigtlaender** [elm-lang/elm-compiler#1359]
Replaces https://github.com/elm-lang/elm-compiler/issues/1223, has SSCCE.

This program:
```elm
import Html

bad param =
  identity <| case param of
    Just _ ->
      0

    Nothing ->
      let
        param = True
      in
        if param then 1 else 2

main = Html.text (toString (bad Nothing))
```
leads to a runtime error.

The 0.17 produced JS is:
```js
var _user$project$Temp1461826735203948$bad = function (param) {
	return _elm_lang$core$Basics$identity(
		function () {
			var _p0 = param;
			if (_p0.ctor === 'Just') {
				return 0;
			} else {
				var param = true;
				return param ? 1 : 2;
			}
		}());
};
```
Why that is wrong is explained in https://github.com/elm-lang/elm-compiler/issues/1223#issuecomment-169645306.

There is also a test case PR open, as https://github.com/elm-lang/elm-compiler/pull/1253.

**Comment by evancz** [comment]
Tracking in #1377

**Comment by jvoigtlaender** [comment]
A related code gen bug, though with slightly different characteristics (no "function call around a case expression") was raised as https://github.com/elm-lang/elm-compiler/issues/1428.

**Comment by splodingsocks** [comment]
Just wanted to add another simple case that produces a runtime error with variable shadowing:

```elm
import Html

foo = Just "abcd"

x = 
  let
    foo = foo
  
    bar = case foo of
      Just thing -> "whatever"
      _ -> "no worries"
  in
    "doesn't matter"
    
    
main = Html.div [] [Html.text "whumf"]
```

In this case, removing the `foo = foo` fixes the runtime error.

**Comment by jvoigtlaender** [comment]
@splodingsocks, no, this is an unrelated error. It has nothing to do with shadowing, as you can easily convince yourself of by playing around with it some more, replacing the `foo = foo` line by `something = something` and/or changing other occurrences of `foo` in consistent ways.

What it is an instance of is #873.

So my suggestion is that you delete your comment above, and I will delete this comment here, so as to not clutter this issue here with unrelated content.

**Comment by jvoigtlaender** [comment]
@splodingsocks, convinced?

**Comment by splodingsocks** [comment]
Yes, sorry @jvoigtlaender! I've been busy and hadn't gotten back to this yet. I'll take care of the comment above. Thank you!

**Comment by brianhempel** [comment]
Perhaps a slightly simpler example. Outputs "undefined":

```elm
import Html

main =
    let x = "outer" in
    (\_ -> 
        let shouldBeOuter = x in
        let x = "inner" in
        Html.text shouldBeOuter
    ) ()
```

```js
var _user$project$Main$main = _elm_lang$virtual_dom$Native_VirtualDom.staticProgram(
	function () {
		var x = 'outer';
		return function (_p0) {
			var shouldBeOuter = x;
			var x = 'inner';
			return _elm_lang$html$Html$text(shouldBeOuter);
		}(
			{ctor: '_Tuple0'});
	}());
```
Javascript closure environments = 😞

**Comment by evancz** [comment]
Variable shadowing like this will be disallowed in 0.19. You can use the same name as things that appear in an `exposing`, but you will not be able to shadow variables within a local definition.
